### DISCUSSION TITLE
Shipping more quickly

### DISCUSSION THREAD

**Discussion by kgodey** [mathesar-foundation/mathesar#1046] [Maintainer discussions]
Our current top priority is to get Mathesar in the hands of real users as soon as possible. In our [weekly discussion this week](https://github.com/centerofci/mathesar/discussions/1033), we talked about how our work has slowed down in the last few weeks. There were a lot of ideas about how we can increase our speed. 

I'll put each idea in its own post to encourage targeted discussion. Others are welcome to add new ideas as well:
- [Adopting a "Feature First" Approach](https://github.com/centerofci/mathesar/discussions/1046#discussioncomment-2106537)
- [Product & Design Review Teams](https://github.com/centerofci/mathesar/discussions/1046#discussioncomment-2106540)
- [Product & Design Creation Process](https://github.com/centerofci/mathesar/discussions/1046#discussioncomment-2106543)
- [Facilitating Synchronous Discussions](https://github.com/centerofci/mathesar/discussions/1046#discussioncomment-2106544)
- [Reducing Notifications](https://github.com/centerofci/mathesar/discussions/1046#discussioncomment-2106547)
- [Speeding up Code Review](https://github.com/centerofci/mathesar/discussions/1046#discussioncomment-2106548)
- [More Focused Issues](https://github.com/centerofci/mathesar/discussions/1046#discussioncomment-2106549)
- [Milestone Scope](https://github.com/centerofci/mathesar/discussions/1046#discussioncomment-2106550)
- [Behavior Driven Development](https://github.com/centerofci/mathesar/discussions/1046#discussioncomment-2106552)

**Comment by kgodey** [comment]
## Adopting a "Feature First" Approach

Many of us have been guilty of focusing on perfecting technical architecture over completing features quickly. Instead, our primary goal should be completing the feature as specced in the design, even if that involves something "hacky" or isn't great for future extensibility. We should also not be over-designing features to work more generally (e.g. aiming towards releasing a general component library or a general release of the `db` backend module), the work we do should support the UX we have planned and no more.

With this approach, we will accrue technical debt, but that's a trade off we're willing to make at this stage. To help mitigate this, it's always helpful to open a GitHub issue to track known issues and hacks when it's fresh on your mind.

Everyone should:
- be extremely vigilant about making sure you're prioritizing feature work over architecture on a day-to-day basis.
- speak up if you could use help with making sure you're focusing on the right things.

**Reply by dmos62** [reply]
Tracking technical debt in the issue tracker is a good idea. Currently we have one technical debt related label `affects: technical debt`. It seems to be meant to label solutions to technical debt. Might be helpful to introduce another label like `type: technical debt` for just tracking occurrences of tech debt.

**Reply by kgodey** [reply]
I'm not sure what the difference between a solution to technical debt and an occurrence of technical debt is.

**Reply by mathemancer** [reply]
A likely sacrifice of this approach will be the accessibility of the code base for outside contributors.  (It will also bog _us_ down over time, but I think that's already in the calculation mentioned above regarding tech debt.)

**Reply by dmos62** [reply]
> I'm not sure what the difference between a solution to technical debt and an occurrence of technical debt is.

It's the difference between a problem and a solution. One might notice that there's a problem that we would want to address, but might not see the obvious correct solution to that problem. If that were the case, the hypothetical poster couldn't formulate the ticket as a solution.

**Reply by kgodey** [reply]
@dmos62 I think it's fine to have the "Proposed solution" section in the issue template say something like "Not sure, need to research and evaluate solutions." and set the status label of the ticket to `status: draft`. 

I'm also fine with changing the `affects: technical debt` label to `type: technical debt`, but I don't think we should have two labels for technical debt since it will be confusing to figure out what to use when.

**Comment by kgodey** [comment]
## Product & Design Review Teams

@seancolsen has suggested creating dedicated product and design review teams to review new product and design work. I think this is a good idea, it will reduce the number of people involved in the review process and make responsibility for who needs to review each spec very clear.

@ghislaineguerin and @kgodey will be on both teams. We should have at least one backend and one frontend engineer on both as well. **We need volunteers for the product and design review teams, please post below.**

Any other thoughts on this proposal are welcome here as well.

**Reply by seancolsen** [reply]
I'd like to be on the Design Team.

**Reply by pavish** [reply]
I'd like to be on both teams.

**Reply by silentninja** [reply]
I'd like to be on the Product Team

**Reply by mathemancer** [reply]
I'd like to be on both teams.

**Reply by dmos62** [reply]
I think these reviews will take some time, so I can be on the product team to share the workload.

**Reply by kgodey** [reply]
I've documented the Design Review team on the [Design Review Guidelines](https://wiki.mathesar.org/en/design/process/review-guidelines) page.

There is no wiki page describing the product review process yet, it's been pretty ad-hoc so far. Once we've finalized how the process works, I'll document the product review team in the relevant documentation.

**Comment by kgodey** [comment]
## Product & Design Creation Process

No changes are planned here since I don't think it is affecting our speed, product and design is well ahead of implementation at the moment. 

One of the things that came up is that visual design should be improvised during implementation. I don't think this is the case - visual design should follow the prototype as closely as possible and only be improvised if there's something missing in the prototype. @ghislaineguerin and I are putting a lot of thought into design patterns and that should be reflected in the actual UI.

**Reply by seancolsen** [reply]
> visual design should follow the prototype as closely as possible and only be improvised if there's something missing in the prototype

I find this statement to be at odds with my experience here thus far. I can elaborate, but I think it be best to discuss on a sync call with @kgodey @ghislaineguerin @pavish. Perhaps we can add this to the agenda for our next sync meeting? Back end people could skip this topic I think.

**Reply by ghislaineguerin** [reply]
I agree on having a sync call to discuss this

**Reply by seancolsen** [reply]
> I don't think it is affecting our speed

For me personally, product and design has been the `#2` thing slowing _me_ down. (Second to the over-prioritization of architecture perfection).

Why?

- The product and design process has been slowing my productivity because I've spent significant time participating in product and design _review_ that I could have spent building front end features. I've recently decided to personally de-prioritize my involvement in product and design review, but I think it's worth considering our workflow as we onboard new team members so that they don't get sucked into participating in the product and design review more than they should. The new teams structure mentioned above will help a lot I think!
- (_more minor note_) The features I've worked on have often required significant improvisation during implementation, sometimes opening up new design questions. Most recently #1005, for example. Personally, I'd be able to move faster if the designs were more specific. I'm used to working with designs that specify the UI in near-pixel-perfect terms. However, creating more specific designs might not actually improve our overall team efficiency, as it would just be shifting some of the design work earlier in the process.

**Reply by seancolsen** [reply]
> I agree on having a sync call to discuss this

I added it to the [agenda](https://hackmd.io/OEXWY519TceOUi53q0Ay1A)

**Reply by kgodey** [reply]
@seancolsen By "product and design process" in this point, I mean only the spec and design creation process. The review process is addressed under [Product and Design Teams](https://github.com/centerofci/mathesar/discussions/1046#discussioncomment-2106540) above, hopefully that helps.

I agree it makes sense to discuss how much improvisation/design specificity is necessary at the next weekly meeting.

**Reply by seancolsen** [reply]
Thanks for the clarification. Yes I'm in agreement with you now, @kgodey

**Comment by kgodey** [comment]
## Facilitating Synchronous Discussions

I've set up a weekly meeting for 6 weeks to hopefully improve our process around having synchronous discussions to resolve issues. People are encouraged to add agenda items to the meeting notes document any time during the week. We will cancel the meeting if there are no agenda items set in the HackMD meeting notes by the 5 PM US Eastern time on the previous day.

This is not meant to be a replacement for more targeted meetings, it's just to help us ease into the habit of talking to each other to resolve issues. We should still be proactively scheduling targeted meetings about specific issues if they will be helpful.

**Comment by kgodey** [comment]
## Reducing Notifications

We discussed this during the State of Mathesar meeting so I won't go into more detail here. Just a couple of questions:
- How does everyone feel about the level of notifications that you're getting? Do you need help reducing them?
- Do people find the `Code - Review` channel useful or should we try to use GitHub notifications exclusively for review?

**Reply by seancolsen** [reply]
I'm in favor of eliminating `Code - Review`.

**Reply by pavish** [reply]
Our `Code - Review` channel is currently also used for other purposes. On the top of my head:
* Indicating urgency of certain PRs
* Asking optional reviews on draft PRs
* Informing everyone on specific changes, without needing their input
* Informing updates on our staging repo 

Assigning people on GH for these cases have a larger round-trip time, which may not be desirable.

I personally don't find the notification level to be distracting.

Having said that, I think we can use the `Code - frontend/backend` channels for these scenarios, and eliminate `Code - Review`.

**Reply by dmos62** [reply]
> How does everyone feel about the level of notifications that you're getting? Do you need help reducing them?

My notification information-to-noise ratio is good. However, it would be better if I didn't get automatically subscribed to issues I wasn't mentioned or requested in. I think it was mentioned that I can configure that; haven't taken the time yet (I guess it's not that bad yet.

> Do people find the Code - Review channel useful or should we try to use GitHub notifications exclusively for review?

I do. I feel more assured that my review requests are being noticed without unnecessary delay. I can also easily track others' PRs' progress.

**Reply by seancolsen** [reply]
I've left the `Code - Review` channel. I imagine that should cover my concerns, as long as people are also tagging me in GH when needed.

**Reply by kgodey** [reply]
> Having said that, I think we can use the `Code - frontend/backend` channels for these scenarios, and eliminate `Code - Review`.

We were doing this in the general code channel before we had a review channel but we decided to create a dedicated channel because review requests would often get drowned by longer conversations. I think we should keep the `Code - Review` channel and people who find the notifications distracting can leave or mute. As long as PRs are getting reviewed in a timely manner, how people choose to get notified is fine.

@seancolsen I assume you'll temporarily join the `Code - Review` channel in case you need to update others about the types of things @pavish mentioned.

**Reply by kgodey** [reply]
I updated the [Code Review Guidelines](https://wiki.mathesar.org/en/engineering/code-review) wiki page to say that posting on the `Code - Review` channel is optional and can be used for urgent PRs or optional review of draft PRs.

**Comment by kgodey** [comment]
## Speeding up Code Review

These points came out of some 1:1 calls I had with various people but I think it's worth mentioning here.

- We should be aiming to turn around PR reviews within 1-2 working days. This means that there should be movement on the PR every 1-2 days, a new review, code review fixes, repeat if needed, and then merge.
- We should be aiming to merge PRs in and create new issues for improvements rather than keeping PRs in review until every possible issue is fixed.
- Code review should be a fairly quick process. Reviewers should be focused on asking the right questions, not on doing research into the answers and suggesting them. 
    - e.g. if you're wondering if the author considered a particular implication of a change they made, ask them that instead of doing research into all the implications yourself and informing the author of them.
- If you'd like to reconsider the architecture of a PR, create a draft issue for figuring that out rather than blocking the PR until you figure out the right architecture.
- When reviewing community contributed PRs, if it's easier to make the changes yourself rather than describe the changes needed as a code review, just make the changes and merge the PR. You can explain what you did and thank the contributor for their work.

**Reply by seancolsen** [reply]
> We should be aiming to turn around PR reviews within 1-2 working days

The 1-2 working days rule-of-thumb seems good for small PRs, and I think we should always be aiming to create small PRs when possible -- particularly because they're easier to review. But for larger PRs (which absolutely are necessary sometimes), I'd extend the rule-of-thumb. I not sure how to communicate a more nuanced rule-of-thumb for longer PRs though. For example I spent 5 days reviewing #915. I can certainly try to trim down my critique going forward. But I expect big PRs to take longer. I'm currently working on reviewing #926 which has also been tough to fit into 2 days.

**Reply by kgodey** [reply]
That's fine, this is meant to be a guideline that helps us make smaller PRs and not overthink code review. I don't think we need a separate rule of thumb for larger PRs.

**Reply by dmos62** [reply]
We could get creative in making PRs smaller. For example, splitting a large PR into smaller PRs that target a non-master branch. That way we can make PRs that break things or don't make sense on their own, provided that upcoming PRs will resolve that, and when ready merge that branch into master.

We did something similar for the current functions API PR, but instead of targeting a dedicated branch, we're targeting master, but structured changes in a way that doesn't affect existing APIs. The breaking changes will come in a subsequent PR.

**Reply by seancolsen** [reply]
@kgodey I found myself revisiting your above bullet points for a second time today, just to help steer me towards the best level of critique for the PR I'm currently reviewing. I like your guidelines a lot. After this discussion has settled, I think it would be good to codify those guidelines on the [Code Review Guidelines](https://wiki.mathesar.org/en/engineering/code-review) wiki page, especially to help new team members later on.

**Reply by kgodey** [reply]
@seancolsen I'll do that, thanks.

**Reply by kgodey** [reply]
These guidelines are now in the [Code Review Guidelines](https://wiki.mathesar.org/en/engineering/code-review) page.

**Comment by kgodey** [comment]
## More Focused Issues

[There was some feedback](https://github.com/centerofci/mathesar/discussions/1033#discussioncomment-2090851) from @silentninja about creating more focused issues, both in terms of separation of concerns and level of detail in each issues. I do think having clear expected outcomes for each issue will help us keep PRs more focused. It will also help new contributors get into issues more easily, which is a nice side benefit.

I think the takeaway here is to put some extra time into creating detailed issues with clear expected outcomes, it will save us time in the implementation phase. @seancolsen has been doing a great job with the new issues for the `NULL` and foreign key specs, and they can serve as a model to emulate.

**Comment by kgodey** [comment]
## Milestone Scope

Feedback from @silentninja:

> Linear Milestone - Currently our milestones are linear, so this makes our work aim for perfection. We should rather aim to have an extended(or future) counterpart for the milestone. While we have it various forms as of now(future tags, future ideas wiki page), it would be much better if we have it as a milestone, as it would make us actually use it and think about it when working on day to day basis,

The scope for each milestone is decided during the design phase – we create the design based on requirements for the alpha release and then create GitHub issues to implement that design. I'm not sure that redoing the way we do milestones will have much effect as long as we're only focused on implementing what we spec.

Further thoughts welcome.

**Reply by seancolsen** [reply]
I think that the "More Focused Issues" intervention (above) will offer some help on avoiding the "aim for perfection" pitfall by allowing us to apply more granular prioritization to features. We can put the important bits in the nearby milestones. And put the less important bits later on down the roadmap.

**Reply by silentninja** [reply]
I don't mean to say we redo the milestone from design spec, rather the extended milestone is for issues that crop up when implementing the features. 

Right now most of the future issues are something we are unsure of or good to have features, rather it should be comprised of features we don't need for `alpha` but will be coming in `beta`, by not having an `extended` milestone, we tend to give off the vibe that there won't be a further enhancement to this existing milestone. Having an `extended/future` milestone would make us think we don't have to do it immediately.

It could also benefit us in future as we wouldn't have to reorganise milestones again.

**Reply by kgodey** [reply]
@silentninja So far, we have not been speccing out features past the alpha release. Could you give me an example of an issue that would go in an extended milestone? Multiple examples would be better.

**Comment by kgodey** [comment]
## Behavior Driven Development

Feedback from @silentninja:

> Behaviour Driven Developement - Having BDD style test for `UI` apis and unittest for  `db` and logic testing of `UI` apis would help us focus better on features rather than architecture, creating a good separation of concern.

I'm not sure if it's worth taking the time to introduce a new type of testing before the alpha release, especially since not everyone on the team is familiar with the tools involved. I think we should defer this until after the alpha release.

Further thoughts welcome.

**Reply by seancolsen** [reply]
I agree with Kriti that it's best not to pursue a new testing pattern right now. The E2E testing has already been a pretty huge time suck, and I've been tempted to say "screw E2E altogether", but it does feel like we're pretty close at this point. I just want to get like a dozen-ish E2E tests locked in and then move on to features, at which point we can continue adding relatively sparse E2E tests.

**Reply by silentninja** [reply]
BDD is more of a paradigm than tooling, we could do it without any tools if required. BDD would force us to think about what is needed from a feature perspective than jumping into implementation. Architecture first tends to come from getting into implementation detail 
instead of trying to get the bare minimum to pass the test and then focusing on making the architecture perfect as needed.
I definitely wouldn't introduce it to the existing testing as it would be really inconsistent and not worth it, but would rather start with `e2e` tests and create `integration` tests required for the `e2e` test. It would make us stay focused on what the requirements are

**Reply by pavish** [reply]
@silentninja I think we already do a fair amount of BDD, though we may not call it that. Our product & design specs define the behaviours from the feature & requirement perspective and then we get into implementation perspective on the issues. 

For testing however, I think by BDD testing, you're referring to how we should describe the tests, and help that description shape our abstractions on the testing layer, thereby directing our code architecture. (Let me know if I'm mistaken).

We could do that with the same tools we use. We currently only have the playwright automation layer for the E2E tests with a pytest runner, we might just need a higher layer for BDD testing description.

While I do think this is a good idea, I don't think it's worth focusing at the moment since not all of us (including me) know the best ways to write a BDD test. I'm all up for jumping deep into this later, probably after our initial release.

**Reply by silentninja** [reply]
@pavish You are right with saying product and design specs are focusing on the behaviour, but the issue occurs when we start implementing features as there is a possibility to start with implementation details rather than focus on the feature first.

By writing these features, it would make it really hard to add architectural tests, forcing you to write a different set of tests, hence forcing you to have granular issues, but all these would work if we go with the `test first` approach.

Pytest already has [pytest-BDD](https://pypi.org/project/pytest-bdd/), which uses a subset of Gherkin language, all it does is add a few descriptive titles on top of test cases. `Jest` allows for such behaviour but vanilla pytest does not. 

We could even write comments skipping any frameworks, but adding frameworks would enforce certain behaviour. 

I do agree we would want to trim things down and not add unnecessary features, but this would enforce certain behaviour and might even be beneficial as a self-documentation of how our features should work.

BDD framework is like a linter that enforces a behaviour.  Moreover, I want to introduce BDD to E2E tests and not to our existing unit tests/integration tests and if does seem like a good candidate and we benefit from it, we can introduce it our existing tests

**Reply by seancolsen** [reply]
@silentninja 

### BDD

Having never before used BDD or Gherkin, I'm interested to learn. I skimmed over the Gherkin page when you first suggested it to me (in think in a DM, if I recall), and my initial reaction was _"thinking more deeply about this does not seem like a good use of my time right now"_. 

> BDD is more of a paradigm than tooling, we could do it without any tools if required.

Cool. If there are ways you could nudge us in the direction of BDD with suggestions that are quick and straightforward, I'd be keen to explore them.

### Page Object Model

You also suggested Page Object Model to me, and I had the same reaction -- I'm trying to get this E2E stuff done as quick as possible so I can get back to building features. I did notice that Playwright has some [features](https://playwright.dev/docs/test-pom) built around Page Object Model. In my web searching to troubleshoot issues with Playwright, I also noticed that a lot of people using Playwright seem to be also use Page Object Model. So I think it might be worth exploring more, once we're doing more E2E testing.

### Next steps

My rough plan with the E2E testing is:

1. Get #1032 ready to merge by satisfying all the goals mentioned in the PR description.
1. Make another PR which adds roughly 10 more E2E tests, providing more coverage.

After that, it might be a good time for you to open a PR which makes some changes to one or more of our E2E tests, adapting them to follow BDD (or even Page Object Model). With BDD being more of a paradigm (and thus more nebulous, conceptually), I think that it will be easier for other team members like me to critique your suggestions if we're looking at something concrete like a diff.

### Terminology

> would rather start with e2e tests and create integration tests required for the e2e test.

I also want to acknowledge that we seem to be using the terms "E2E" tests and "Integration" tests to refer to the same thing -- the _highest_ level of testing -- whereas other teams might use these terms more distinctly, with "E2E" being the highest level and "integration" being slightly lower level (like our current API tests). I'm fine with this though. Testing terminology has always seemed somewhat blurry and subjective to me.

**Reply by silentninja** [reply]
Maintenance work is a necessary evil, From a developer's, perspective testing is mostly a time drain and might seem like not worth it, but it pays off in the long run. It might seem tempting to put architectural work and maintenance work for later, we might be accruing technical debt which we might have to pay off with interest.

 Introducing certain things in future might mean that we would have to redo certain things and the larger its impact, the harder it will be to introduce a change, for example trying to introduce BDD to backend testing would be really hard at this point and I am not sure if it is even worth it. While I do acknowledge features are what the users will be using, we should also be mindful of our technical debt. 

Like @seancolsen said, I could try making a draft PR for the E2E test cases and if does seem plausible, we could use it.

**Reply by kgodey** [reply]
@silentninja The issue with this suggestion is that I'm not sure what problem it solves – you've indicated that it will help us think about features more, but that seems like a high-level point. For most of the other points in this discussion (e.g. speeding up code review, creating more focused issues, creating the product and design review teams), I can come up with a specific recent issue or discussion that triggered the workflow change. Even the current E2E tests that @seancolsen is working on are being prioritized because it's slowing down frontend work to manually test that previous features are not broken while building a new feature. 

Could you give us a concrete example of how you putting extra time into this (which also entails taking away time from our specced features) will help us increase speed *in the short term*?

**Reply by silentninja** [reply]
I think there is a miscommunication here

1. BDD is a way of writing test cases, it does not even need tooling or any extra time, this is just like code styling. Most of the test cases that test the service layer of the backend are similar to BDD, just that it is not explicitly stated

```
Given a table with data
When a filter is applied
Should return the filtered response
```

I am just trying to say that we if follow this method, we start implementation from how to make a feature work and then create the required architecture. This is also possible if we start writing test cases first, just that BDD would give us clarity as it is just plain English. This is to address the concern of [how to make us think in terms of features first](https://github.com/centerofci/mathesar/discussions/1046#discussioncomment-2106537). 

2. BDD tools - While not necessary, but can enforce the behaviour just a good linting tool. This also gives us the added benefit of having a human-readable test case that can serve as good documentation and easier to search test cases.

BDD would be a disaster we tend to implement features and then write test cases just for the sake of it.

@kgodey This issue is to adopt a [Feature first approach](https://github.com/centerofci/mathesar/discussions/1046#discussioncomment-2106537)

**Reply by kgodey** [reply]
@silentninja Thanks for the explanation. I'm still not sure what the action items for us would be, though. You mention that we should "follow this method" - what concrete changes are you proposing that we make to our process, workflow, and/or code?
